Coccinelle improvements

mrc0mmand opened the pull request:
During my another archaeology expedition into the depths of our issue tracker I stumbled upon https://github.com/systemd/systemd/issues/7286 and more particularly @poettering's https://github.com/systemd/systemd/issues/7286#issuecomment-343496314. After a while of playing around with Coccinelle I noticed that `spatch` generates really garbled results for `coccinelle/equals-null.cocci`, which was also mentioned by @poettering in 234519ae6d6821ae5e83bcddb211905ef4de5700.

Apparently, you need to explicitly define statement isomorphisms to make this work as expected. Coccinelle does something very similar, but in the exact opposite direction we'd like to do (i.e. `!X => X == NULL`). So, after writing a short isomorphism file it looks like Coccinelle finally gives reasonable results:

For example, after applying:
```diff
diff --git a/src/core/path.c b/src/core/path.c
index 5a00150e9b..28d1e178a8 100644
--- a/src/core/path.c
+++ b/src/core/path.c
@@ -283,7 +283,7 @@ static int path_verify(Path *p) {
         if (UNIT(p)->load_state != UNIT_LOADED)
                 return 0;
 
-        if (!p->specs) {
+        if (p->specs == NULL) {
                 log_unit_error(UNIT(p), "Path unit lacks path setting. Refusing.");
                 return -ENOEXEC;
         }
@@ -296,7 +296,7 @@ static int path_add_default_dependencies(Path *p) {
 
         assert(p);
 
-        if (!UNIT(p)->default_dependencies)
+        if (UNIT(p)->default_dependencies != NULL)
                 return 0;
 
         r = unit_add_dependency_by_name(UNIT(p), UNIT_BEFORE, SPECIAL_PATHS_TARGET, true, UNIT_DEPENDENCY_DEFAULT);
```

Coccinelle gives following output:
```
$ spatch --iso-file coccinelle/systemd-definitions.iso --sp-file coccinelle/equals-null.cocci src/core/path.c  
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: src/core/path.c
diff = 
--- src/core/path.c
+++ /tmp/cocci-output-6880-c5f492-path.c
@@ -283,7 +283,7 @@ static int path_verify(Path *p) {
         if (UNIT(p)->load_state != UNIT_LOADED)
                 return 0;
 
-        if (p->specs == NULL) {
+        if (!p->specs) {
                 log_unit_error(UNIT(p), "Path unit lacks path setting. Refusing.");
                 return -ENOEXEC;
         }
@@ -296,7 +296,7 @@ static int path_add_default_dependencies
 
         assert(p);
 
-        if (UNIT(p)->default_dependencies != NULL)
+        if (UNIT(p)->default_dependencies)
                 return 0;
 
         r = unit_add_dependency_by_name(UNIT(p), UNIT_BEFORE, SPECIAL_PATHS_TARGET, true, UNIT_DEPENDENCY_DEFAULT);
``` 

If you omit the isomorphism file, it ends up with an error (in this particular case):
```
$ spatch --sp-file coccinelle/equals-null.cocci src/core/path.c  
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: src/core/path.c
     (ONCE) already tagged but only removed, so safe
     
previous modification:
MINUS
  >>> if (e)


According to environment 1:
   rule starting on line 8.e -> UNIT(p)->default_dependencies


current modification:
MINUS
  >>> if (e)


According to environment 1:
   rule starting on line 8.e -> UNIT(p)->default_dependencies != NULL


rule starting on line 8: already tagged token:
C code context
File "src/core/path.c", line 299, column 49, charpos = 8344
  around = ')',
  whole content =         if (UNIT(p)->default_dependencies != NULL)
```

Or without the *synthetic* patch, it gives following nonsensical results:
```
@@ -628,8 +628,9 @@ static int path_deserialize_item(Unit *u
                 f = path_result_from_string(value);
                 if (f < 0)
                         log_unit_debug(u, "Failed to parse result value: %s", value);
-                else if (f != PATH_SUCCESS)
-                        p->result = f;
+                else {if (f != PATH_SUCCESS)
+                                p->result = f;
+                }
 
         } else
                 log_unit_debug(u, "Unknown serialization key: %s", key);
@@ -666,8 +667,9 @@ static int path_dispatch_io(sd_event_sou
         /* log_debug("inotify wakeup on %s.", u->id); */
 
         LIST_FOREACH(spec, s, p->specs)
-                if (path_spec_owns_inotify_fd(s, fd))
+        {if (path_spec_owns_inotify_fd(s, fd))
                         break;
+        }
 
         if (!s) {
                 log_error("Got event on unknown fd.");
```

@poettering: if there's still interest in making this work, I could take a look at the other scripts, to reduce the amount of false-positives to minimum

mrc0mmand commented:
The `take-fd.cocci` patch fixes this particular issue:

```
$ spatch --iso-file coccinelle/systemd-definitions.iso --sp-file coccinelle/take-fd.cocci src/basic/errno-util.h 
init_defs_builtins: /usr/lib64/coccinelle/standard.h
HANDLING: src/basic/errno-util.h
diff = 
--- src/basic/errno-util.h
+++ /tmp/cocci-output-28263-971baa-errno-util.h
@@ -15,8 +15,7 @@ static inline void _reset_errno_(int *sa
 
 #define UNPROTECT_ERRNO                         \
         do {                                    \
-                errno = _saved_errno_;          \
-                _saved_errno_ = -1;             \
+                errno = TAKE_FD(_saved_errno_);             \
         } while (false)
 
 static inline int negative_errno(void) {
```

This is clearly wrong, as `errno` is not a file descriptor. Ideally, this should be fixed by using `local idexpression`, however, that wouldn't match, for example, "x->fd", as it's considered an expression by the SmPL language[0]. Fixing it by explicitly stating `expression p != errno;` is not particularly nice, but it should serve its purpose.

[0] Quote from http://coccinelle.lip6.fr/docs/main_grammar.pdf
> A metavariable declaration local idexpression v means that v is restricted to be a local variable. If it should just be a variable, but not necessarily a local one, then drop local. A more complex description of a location, such as a->b is considered to be an expression, not an idexpression.

keszybz commented:
The changes look reasonable, but can you please update the commits to have some description? Probably just pasting some of the text of the PR would be enough. The intent and effect of those changes is not obvious...

mrc0mmand commented:
> The changes look reasonable, but can you please update the commits to have some description? Probably just pasting some of the text of the PR would be enough. The intent and effect of those changes is not obvious...

Definitely. I wasn't sure if we want to pursue this, so I didn't bother with commit descriptions. Should be fixed now.

mrc0mmand commented:
Also, @keszybz, could you, by any chance, try running Coccinelle with this PR applied and report any discrepancies? The current output looked alright to me, but I'd be glad if someone with a proper C knowledge would take a look as well, so we can fix it in one batch.

keszybz commented:
`coccinelle/run-coccinelle.sh -i coccinelle/equals-null.cocci` gives a long list of changes, and they look correct to me.

`coccinelle/run-coccinelle.sh -i coccinelle/take-fd.cocci` does no changes.

keszybz commented:
Please push a commit with the changes too. Let's merge everything together in one PR.

mrc0mmand commented:
Will do. I'd also like to go through the other transformation files and check if they generate correct outputs, and if so, apply those changes as well. If not, try to fix them, like I did in 906c167a50e5080e3bb49a53ad63dd7c926fbd3e. (Stop me if it's too much)

mrc0mmand commented:
Soo, I tried to fix as much false positives as I could and here's the results.

I added an exclusion list to the `coccinelle/run-coccinelle.sh` script, as there are some paths which shouldn't be *touched* by Coccinelle, namely
* `src/basic/linux/*`
* `src/shared/linux/*`
* `*src/boot/efi/*`

I also added few other patches for various transformation files to filter out nonsensical transformations. After that, I run all Coccinelle scripts and the results can be found in the last commit of this PR. PTAL!

keszybz commented on coccinelle/const-strlen.cocci:
```diff
@@ -1,8 +1,12 @@
 @@
+(
+#define STRLEN
+&
```
Nice.

keszybz commented on src/coredump/coredump.c:
```diff
@@ -800,7 +800,7 @@ static int submit_coredump(
                  * The target was set previously to something safe. */
                 assert(startswith(core_message, "MESSAGE="));
-                log_dispatch(LOG_ERR, 0, core_message + strlen("MESSAGE="));
+                log_dispatch(LOG_ERR, 0, core_message + STRLEN("MESSAGE="));
```
Please drop this patch. STRLEN only makes sense in very narrow circumstances: when the length is used to define a VLA. By using STREN() we avoid VLA creation and get a normal stack allocation. In all other circumstances, the compiler replaces strlen on a constant by a constant integer, even at `-O0`.

keszybz commented on src/import/importd.c:
```diff
@@ -250,7 +250,7 @@ static void transfer_send_logs(Transfer *t, bool flush) {
 
                 /* Skip over NUL and newlines */
-                while (e < t->log_message + t->log_message_size && (*e == 0 || *e == '\n'))
+                while (e < t->log_message + t->log_message_size && (IN_SET(*e, 0, '\n')))
```
Parens are not needed here now.

keszybz commented on src/shared/bootspec.c:
```diff
@@ -591,8 +591,8 @@ static int boot_entries_select_default(const BootConfig *config) {
-                return -1; /* -1 means "no default" */
+                return log_debug_errno(SYNTHETIC_ERRNO(1),
+                                       "Found no default boot entry :("); /* -1 means "no default" */
         }
```
Parens can be dropped.

keszybz replied:
This is very ugly. -1 is not an errno code but an index here. I'd prefer to leave this as it was. Sorry for not noticing this earlier.

keszybz commented on src/shared/securebits-util.h:
```diff
@@ -7,7 +7,7 @@ int secure_bits_to_string_alloc(int i, char **s);
 
 static inline bool secure_bits_is_valid(int i) {
-        return ((SECURE_ALL_BITS | SECURE_ALL_LOCKS) & i) == i;
+        return FLAGS_SET(SECURE_ALL_BITS | SECURE_ALL_LOCKS, i);
```
Technically, this is valid. But it's super misleading, because the flags are here in the position of the value. Maybe skip this one?

keszybz commented on src/test/test-string-util.c:
```diff
@@ -314,7 +314,7 @@ static void test_strshorten(void) {
         assert_se(strlen(strshorten(s, 12)) == 6);
         assert_se(strlen(strshorten(s, 2)) == 2);
-        assert_se(strlen(strshorten(s, 0)) == 0);
+        assert_se(isempty(strshorten(s, 0)));
```
Please drop this, this is testing code for `strshorten` and we don't want to mix `isempty` into it.

keszybz commented on src/test/test-string-util.c:
```diff
@@ -540,8 +540,8 @@ static void test_first_word(void) {
-        assert_se(strlen_ptr("") == 0);
-        assert_se(strlen_ptr(NULL) == 0);
+        assert_se(isempty(""));
+        assert_se(isempty(NULL));
```
Same ehre.

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -203,8 +203,8 @@ static int cd_capability_compat(int fd) {
-                return -1;
+                return log_debug_errno(SYNTHETIC_ERRNO(1),
+                                       "CDROM_GET_CAPABILITY failed");
         }
```
Drop parens. Also, don't return the fake value, please just use `errno`.

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -226,8 +226,8 @@ static int cd_capability_compat(int fd) {
-                log_debug("CDROM_DRIVE_STATUS != CDS_DISC_OK");
-                return -1;
+                return log_debug_errno(SYNTHETIC_ERRNO(1),
+                                       "CDROM_DRIVE_STATUS != CDS_DISC_OK");
```
Same here.

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -249,8 +249,7 @@ static int cd_inquiry(int fd) {
-                log_debug("not an MMC unit");
-                return -1;
+                return log_debug_errno(SYNTHETIC_ERRNO(1), "not an MMC unit");
         }
```
Here too, and below.

mrc0mmand replied:
Hmm, thinking about this specific case, I'd say `errno` would have a pretty irrelevant value here. Shouldn't we use `SYNTHETIC_ERRNO(EINVAL)` instead?

keszybz replied:
You're right, `errno` makes not sense here.

keszybz commented on src/udev/scsi_id/scsi_id.c:
```diff
@@ -347,16 +347,18 @@ static int set_options(int argc, char **argv,
                         else {
-                                log_error("Unknown page code '%s'", optarg);
-                                return -1;
+                                return log_error_errno(SYNTHETIC_ERRNO(1),
```
Parens, errno...

mrc0mmand commented:
Thanks for the review! I went through the comments and fixed found issues (and further restricted related transformations, if appropriate). I'm not 100% sure, where to return `-1` or `errno`, so PTAL.

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -474,8 +471,7 @@ static int cd_profiles_old_mmc(int fd) {
                 } else {
-                        log_debug("no current profile, assuming no media");
-                        return -1;
+                        return log_debug_errno(-1, "no current profile, assuming no media");
```
This should be `log_debug_errno(SYNTHETIC_ERRNO(ENOMEDIUM), ...)`. The rules are:
- don't use fake errno values like -1, but the best approximate negative errno code
- if the error value x is received from further up the call stack, use log_debug_errno(x). If it is generated locally for convenience of using log_..._errno(), wrap it in SYNTHETIC_ERRNO().
- there's some code which doesn't follow those rules, it should be made to conform when updating.
- the argument to log_debug_errno() should not have an explicit negative value, i.e. log_debug_errno(r, ...) is OK, but log_debug_errno(-EFOO, ...) should be replaced by log_debug_errno(EFOO, ...).

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -660,8 +656,7 @@ static int cd_media_info(int fd) {
                         if (len & 7 || len < 16) {
-                                log_debug("invalid format capacities length");
-                                return -1;
+                                return log_debug_errno(-1, "invalid format capacities length");
```
Please use `SYNTHETIC_ERRNO(EINVAL)`.

keszybz commented on src/udev/cdrom_id/cdrom_id.c:
```diff
@@ -680,8 +675,7 @@ static int cd_media_info(int fd) {
                                 cd_media = 0; //return no media
-                                log_debug("format capacities returned no media");
-                                return -1;
+                                return log_debug_errno(-1, "format capacities returned no media");
```
Please use `SYNTHETIC_ERRNO(EINVAL)` or `SYNTHETIC_ERRNO(ENOMEDIUM)` or smth.

poettering commented on src/test/test-fd-util.c:
```diff
@@ -155,7 +155,7 @@ static void test_fd_move_above_stdio(void) {
         assert_se(new_fd >= 3);
 
-        assert_se(dup(original_stdin) == 0);
+        assert_se(fcntl(original_stdin, F_DUPFD, 3) >= 3);
```
this should remain dup() actuall. it's one of the few cases where we actually want to see that fd 0 was closed, and thus want plain dup() to fill in fd 0, and not fd 3

poettering commented on src/shared/bootspec.c:
```diff
@@ -591,8 +591,7 @@ static int boot_entries_select_default(const BootConfig *config) {
         if (config->n_entries == 0) {
-                log_debug("Found no default boot entry :(");
-                return -1; /* -1 means "no default" */
+                return log_debug_errno(-1, "Found no default boot entry :("); /* -1 means "no default" */
```
this should remain a regular "return -1". -1 is not really an error here, just a special marker. if we pass it to log_debug_errno() it would be added as ERRNO= field to structured logging if structured logging is used. And we really don't want that.

poettering commented on src/udev/scsi_id/scsi_id.c:
```diff
@@ -347,16 +347,18 @@ static int set_options(int argc, char **argv,
                         else {
-                                log_error("Unknown page code '%s'", optarg);
-                                return -1;
+                                return log_error_errno(-1,
```
SYNTHETIC_ERRNO(EINVAL) or SYNTHETIC_ERRNO(EOPNOTSUPP)?

poettering commented on src/udev/scsi_id/scsi_serial.c:
```diff
@@ -169,8 +169,9 @@ static int scsi_dump_sense(struct scsi_id_device *dev_scsi,
         if (sb_len < 1) {
-                log_debug("%s: sense buffer empty", dev_scsi->kernel);
-                return -1;
+                return log_debug_errno(-1,
```
SYNTHETIC_ERRNO(EINVAL)?

poettering commented on src/udev/scsi_id/scsi_serial.c:
```diff
@@ -182,19 +183,21 @@ static int scsi_dump_sense(struct scsi_id_device *dev_scsi,
-                        log_debug("%s: sense buffer too small %d bytes, %d bytes too short",
-                                  dev_scsi->kernel, sb_len, s - sb_len);
-                        return -1;
+                        return log_debug_errno(-1,
```
...

mrc0mmand commented:
@keszybz, @poettering thanks a lot for the comments, much appreciated, the `SYNTHETIC_ERRNO` stuff is now much clearer to me (and it was one of the reasons why I didn't want to touch the C codebase in the first place). Pushed a new version, PTAL!

keszybz commented on src/udev/scsi_id/scsi_serial.c:
```diff
@@ -182,19 +183,21 @@ static int scsi_dump_sense(struct scsi_id_device *dev_scsi,
-                                return -1;
+                                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "%s: sense result too"
+                                                       " small %d bytes",
```
Please don't split strings like that. (My solution is to patch `spatch` to bump the string length limit to something very large and use that binary instead... But since it's just one case here, it should be enough to patch it by hand ;)) ).

mrc0mmand replied:
I somehow managed to completely miss this one, thanks for noticing!

keszybz commented:
LGTM.

poettering commented:
lgtm to me too.

thanks for cleaning this up! this is very much appreciated. I admit though that I don't grok coccinelle well enough to fully understand the isomorphism stuff...
